From 04ff19c0fe9b576c772d63373f011a6403ec94a3 Mon Sep 17 00:00:00 2001
From: Jason Wang <jasowang@redhat.com>
Date: Fri, 24 Oct 2014 10:57:40 -0400
Subject: [virt] hyperv: Fix the total_data_buflen in send path

Message-id: <1414148260-34692-5-git-send-email-jasowang@redhat.com>
Patchwork-id: 98428
O-Subject: [RHEL7.1 PATCH 4/4] hyperv: Fix the total_data_buflen in send path
Bugzilla: 1156305
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Vitaly Kuznetsov <vkuznets@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

From: Haiyang Zhang <haiyangz@microsoft.com>

Bugzilla: 1156305

total_data_buflen is used by netvsc_send() to decide if a packet can be put
into send buffer. It should also include the size of RNDIS message before the
Ethernet frame. Otherwise, a messge with total size bigger than send_section_size
may be copied into the send buffer, and cause data corruption.

[Request to include this patch to the Stable branches]

Signed-off-by: Haiyang Zhang <haiyangz@microsoft.com>
Reviewed-by: K. Y. Srinivasan <kys@microsoft.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 942396b01989d54977120f3625e5ba31afe7a75c
 from net.git)

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/hyperv/netvsc_drv.c b/drivers/net/hyperv/netvsc_drv.c
index ef0bbd7..a7b6c09 100644
--- a/drivers/net/hyperv/netvsc_drv.c
+++ b/drivers/net/hyperv/netvsc_drv.c
@@ -550,6 +550,7 @@ do_lso:
 do_send:
  /* Start filling in the page buffers with the rndis hdr */
  rndis_msg->msg_len += rndis_msg_size;
+ packet->total_data_buflen = rndis_msg->msg_len;
  packet->page_buf_cnt = init_page_array(rndis_msg, rndis_msg_size,
      skb, &packet->page_buf[0]);
 
-- 
1.7.1